### PR TITLE
add 'dfu' Update Module

### PR DESCRIPTION
**Pull request by mirzak**
The DFU Update Module is a reference Update Module that demonstrates
how to update firmware of a peripheral device using the
Device Firmware Update USB device class.

Changelog: Title

Signed-off-by: Mirza Krak <mirza.krak@northern.tech>

### CODE REVIEW

File: dfu/README.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
+## Description
+
+The DFU Update Module is to able to update peripheral devices connected to the device running Mender.
```
Comment (lluiscampos) [inline dfu/README.md]: is able to update

File: dfu/README.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+```
+wget https://raw.githubusercontent.com/mendersoftware/mender-update-modules/master/dfu/module-artifact-gen/dfu-artifact-gen
+```
+
+Make it executable by running,
+
+```
+chmod +x dfu-artifact-gen
+```
```
Comment (lluiscampos) [inline dfu/README.md]: These could be a single command, same wayas we do for the Update Module itself, `wget ... && chmod +x ...`

File: dfu/module-artifact-gen/dfu-artifact-gen [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+output_path="${artifact_name}.mender"
+
+if [ -z "${device_types}" ]; then
+  echo "Device type not specified. Aborting."
+  show_help_and_exit_error
+fi
+
+if [ -n "${dfu_args}" ]; then
+  eval "dfu_args" > $meta_data_file
+fi
```
Comment (lluiscampos) [inline dfu/module-artifact-gen/dfu-artifact-gen]: `meta_data_file` (and therefore `dfu_args`) seems to be unused

File: dfu/module/dfu [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    NeedsArtifactReboot)
+        echo "No"
+    ;;
+
+    SupportsRollback)
+        echo "No"
+    ;;
+
+    ArtifactInstall)
+        firmware_file=$(ls $FILES/files/)
```
Comment (lluiscampos) [inline dfu/module/dfu]: You could use `firmware_path` here.

File: dfu/module-artifact-gen/dfu-artifact-gen (hunk 3 above)
Comment (lluiscampos) [reply dfu/module-artifact-gen/dfu-artifact-gen]: It is also missing from the help text, but I guess that they are leftovers and you will remove it.

File: dfu/module/dfu [hunk 5]
```diff
@@ -0,0 +1,45 @@
+#!/bin/sh
+
+# This module is useful if you have a peripheral device connected to a running
+# Linux system which is already Linux and Mender and if the device supports
+# firmware updates trough the Device Firmware Update (DFU) device class.
```
Comment (lluiscampos) [inline dfu/module/dfu]: The text is a bit strange: "which is already Linux and Mender [integrated?]"

Comment (lluiscampos) [reply dfu/module/dfu]: Also typo s/trough/through

**Comment by mirzak** [comment]
Thanks for the quick review. Pushed an update, should have handled all your comments.

**Pull request merged**